drv/medusa-seq-server/src/front_io.rs

```diff
@@ -0,0 +1,111 @@
+    ) -> Self {
+        Self {
+            controllers: [
+                FrontIOController::new(fpga_task, 0),
+                FrontIOController::new(fpga_task, 1),
+            ],
+            fpga_task,
+            auxflash_task,
+        }
+    }
+
+    pub fn phy(&self) -> PhySmi {
+        PhySmi::new(self.fpga_task)
+    }
+
+    pub fn present(i2c_task: userlib::TaskId) -> bool {
+        let fruid = i2c_config::devices::at24csw080_front_io0(i2c_task)[0];
+        At24Csw080::validate(&fruid).unwrap_or(false)
+    }
+
+    pub fn initialized(&self) -> bool {
+        self.controllers.iter().all(|c| c.ready().unwrap_or(false))
+    }
+
+    pub fn init(&mut self) -> Result<bool, FpgaError> {
+        let mut controllers_ready = true;
+
+        for (i, controller) in self.controllers.iter_mut().enumerate() {
+            let state = controller.await_fpga_ready(25)?;
+            let mut ident;
+            let mut ident_valid = false;
+            let mut checksum;
+            let mut checksum_valid = false;
+
+            if state == DeviceState::RunningUserDesign {
+                (ident, ident_valid) = controller.ident_valid()?;
+                ringbuf_entry!(Trace::FrontIOControllerIdent {
+                    fpga_id: i,
+                    ident
+                });
+
+                (checksum, checksum_valid) = controller.checksum_valid()?;
+                ringbuf_entry!(Trace::FrontIOControllerChecksum {
+                    fpga_id: i,
+                    checksum,
+                    expected: FrontIOController::short_checksum(),
+                });
+
+                if !ident_valid || !checksum_valid {
+                    // Attempt to correct the invalid IDENT by reloading the
+                    // bitstream.
+                    controller.fpga_reset()?;
+                }
+            }
+
+            if ident_valid && checksum_valid {
+                ringbuf_entry!(Trace::SkipLoadingFrontIOControllerBitstream {
+                    fpga_id: i
+                });
+            } else {
+                ringbuf_entry!(Trace::LoadingFrontIOControllerBitstream {
+                    fpga_id: i
+                });
+
+                if let Err(e) = controller.load_bitstream(self.auxflash_task) {
+                    ringbuf_entry!(Trace::FpgaBitstreamError(u32::from(e)));
+                    return Err(e);
+                }
+
+                (ident, ident_valid) = controller.ident_valid()?;
+                ringbuf_entry!(Trace::FrontIOControllerIdent {
+                    fpga_id: i,
+                    ident
+                });
+
+                controller.write_checksum()?;
+                (checksum, checksum_valid) = controller.checksum_valid()?;
+                ringbuf_entry!(Trace::FrontIOControllerChecksum {
+                    fpga_id: i,
+                    checksum,
+                    expected: FrontIOController::short_checksum(),
+                });
+            }
```

Comment:
style nit, take it or leave it: i  _might_ consider using `continue` here so that the rest of the code doesn't drift to the right; others may disagree with me on this though:
```suggestion
            if ident_valid && checksum_valid {
                ringbuf_entry!(Trace::SkipLoadingFrontIOControllerBitstream {
                    fpga_id: i
                });
                continue;
            }
  
            ringbuf_entry!(Trace::LoadingFrontIOControllerBitstream {
                fpga_id: i
            });

            if let Err(e) = controller.load_bitstream(self.auxflash_task) {
                ringbuf_entry!(Trace::FpgaBitstreamError(u32::from(e)));
                return Err(e);
            }

            (ident, ident_valid) = controller.ident_valid()?;
            ringbuf_entry!(Trace::FrontIOControllerIdent {
                fpga_id: i,
                ident
            });

            controller.write_checksum()?;
            (checksum, checksum_valid) = controller.checksum_valid()?;
            ringbuf_entry!(Trace::FrontIOControllerChecksum {
                fpga_id: i,
                checksum,
                expected: FrontIOController::short_checksum(),
            });
```

Reply:
I hear you on this, but I'm going to opt to keep it the way it is right now. This file is just copy/paste from https://github.com/oxidecomputer/hubris/blob/master/drv/sidecar-mainboard-controller/src/front_io.rs to get a quick solution here. In #1805, which I've not attempted to land before I go on leave, this could get addressed properly.